rbd: check for clusterid mapping in genVolFromVolumeOptions()

Rakshith-R opened the pull request:
This commit adds capability to genVolFromVolumeOptions() to fetch
mapped clusted-id & mon-ips for mirrored PVC on secondary cluster
which may have different cluster-id.

This is required for NodeStageVolume().

We also don't need to check for mapping during volume create requests,
so it can be disabled by passing a bool checkClusterIDMapping as false.


Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: check for clusterid mapping in genVolFromVolumeOptions()

This commit adds capability to genVolFromVolumeOptions() to fetch
mapped clusted-id & mon ips for mirrored PVC on secondary cluster
which may have different cluster-id.

This is required for NodeStageVolume().

We also don't need to check for mapping during volume create requests,
so it can be disabled by passing a bool checkClusterIDMapping as false.

GetMonsAndClusterID() is modified to accept bool checkClusterIDMapping
based on which clustermapping is checked to fetch mapped cluster-id and
mon-ips.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
/retest ci/centos/upgrade-tests-cephfs

Rakshith-R commented:
> /retest ci/centos/upgrade-tests-cephfs

`Sep  6 11:34:37.762: FAIL: failed to create user cephcsi-cephfs-provisioner with error etcdserver: request timed out`
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/upgrade-tests-cephfs/runs/3000/nodes/91/log/?start=0

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -1171,11 +1171,21 @@ func genVolFromVolumeOptions(
-	rbdVol.Monitors, rbdVol.ClusterID, err = util.GetMonsAndClusterID(volOptions)
-	if err != nil {
-		log.ErrorLog(ctx, "failed getting mons (%s)", err)
+	if checkClusterIDMapping {
```
@Rakshith-R just wondering why we need this hard check in place?  ie in createVolume we make it as `false` and in `nodeserver` we make it as `true` .

Rakshith-R replied:
@humblec ,without this check, we may end up using mapped clusterid during create volume request, since we're using this function in parsecreatevolreq too.

humblec reviewed (changes requested):
Can we take very required args as part of  getMonsAndClusterID()?

humblec commented on internal/util/csiconfig.go:
```diff
@@ -131,11 +132,19 @@ func CephFSSubvolumeGroup(pathToConfig, clusterID string) (string, error) {
 // GetMonsAndClusterID returns monitors and clusterID information read from
 // configfile.
-func GetMonsAndClusterID(options map[string]string) (string, string, error) {
+func GetMonsAndClusterID(ctx context.Context, options map[string]string, checkClusterIDMapping bool) (string, string, error) {
```
if you pass `clusterID` we dont need `options` here.

Rakshith-R committed:
util: modify GetMonsAndClusterID() to take clusterID instead of options

This commit:
- modifies GetMonsAndClusterID() to take clusterID instead of options.
- moves out validation of clusterID is set or not out of GetMonsAndClusterID().
- defines ErrClusterIDNotSet new error for reusability.
- add GetClusterID() to obtain clusterID from options.

Signed-off-by: Rakshith R <rar@redhat.com>

humblec commented on internal/util/errors.go:
```diff
@@ -33,6 +33,8 @@ var (
 	ErrSnapNameConflict = errors.New("snapshot name conflict")
 	// ErrPoolNotFound is returned when pool is not found.
 	ErrPoolNotFound = errors.New("pool not found")
+	// ErrClusterIDNotSet is returned when cluster id is not set.s
```
nit: typo , `..id is not set.`

Madhu-1 commented on internal/util/csiconfig.go:
```diff
@@ -131,10 +131,14 @@ func CephFSSubvolumeGroup(pathToConfig, clusterID string) (string, error) {
 // GetMonsAndClusterID returns monitors and clusterID information read from
 // configfile.
-func GetMonsAndClusterID(options map[string]string) (string, string, error) {
-	clusterID, ok := options["clusterID"]
```
why this check is moved out of `GetMonsAndClusterID` and done at multiple places before calling this function. IMO this is error-prone and the user needs to be aware that he/she has to do this check before calling `GetMonsAndClusterID` and it also increases the code duplication and code complexity as one more if check is needed for the caller function.

Rakshith-R replied:
On another note,  we unnecessarily have to create a options map when using this in other places which also leads to code duplication. 

When we have the parameter as clusterID string, it is more obvious to have cluster ID exists check if it's being read from a map. 
@Madhu-1 ^^

Madhu-1 replied:
in that case, please move the validation to a new helper function or define const of `"clusterID"` map key

humblec replied:
I think one thing we can do here is this:

createFunctions to do one thing:

```
 GetMons(clusterID,..)
  just fetch the mons
GetClusterID(options,..)
 just validate the clusterID ..
Check/GetClusterIDMapping()??
```

Rakshith-R replied:
@humblec @Madhu-1  PTAL and let me know if these changes are fine. :)

Madhu-1 reviewed (dismissed):
Hope you have already tested mirror functionality to avoid regressions.

Rakshith-R commented:
> Hope you have already tested mirror functionality to avoid regressions.

If the current changes are fine, I'll do some final tests with different cluster-ids. 
@humblec , does the changes look okay from your end?

humblec commented:
> > Hope you have already tested mirror functionality to avoid regressions.
> 
> If the current changes are fine, I'll do some final tests with different cluster-ids.
> @humblec , does the changes look okay from your end?

Good enough @Rakshith-R , not sure why this has been moved to `draft` though.

Rakshith-R commented:
Current canary image with secondary cluster id `rook-ceph1`
```
[rakshith@fedora rbd]$ k get configmaps rook-ceph-csi-config -o yaml
apiVersion: v1
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph1","monitors":["192.168.61.34:6789"]}]'
  
  [rakshith@fedora rbd]$ k get configmaps rook-ceph-csi-mapping-config -o yaml
apiVersion: v1
data:
  csi-mapping-config-json: '[{"ClusterIDMapping":{"rook-ceph":"rook-ceph1"},"RBDPoolIDMapping":[{"1":"1"}]}]'
```

```
1157-11ec-867a-0242ac110005 GRPC request: {"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-05d02323-5e89-4a50-8893-c4ff5840912f/globalmount","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"clusterID":"rook-ceph","csi.storage.k8s.io/pv/name":"pvc-05d02323-5e89-4a50-8893-c4ff5840912f","csi.storage.k8s.io/pvc/name":"rbd-pvc","csi.storage.k8s.io/pvc/namespace":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-3879d00b-1157-11ec-867a-0242ac110005","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1631180988965-8081-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005"}
E0909 11:09:41.456864  112271 rbd_util.go:1089] ID: 24 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 failed getting mons (failed to fetch monitor list using clusterID (rook-ceph): missing configuration for cluster ID "rook-ceph")
E0909 11:09:41.456899  112271 utils.go:185] ID: 24 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 GRPC error: rpc error: code = Internal desc = failed to fetch monitor list using clusterID (rook-ceph): missing configuration for cluster ID "rook-ceph"

```

With this pr 
```I0909 11:16:07.743167  140126 utils.go:177] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 GRPC call: /csi.v1.Node/NodeStageVolume
I0909 11:16:07.743398  140126 utils.go:181] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 GRPC request: {"secrets":"***stripped***","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-05d02323-5e89-4a50-8893-c4ff5840912f/globalmount","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"clusterID":"rook-ceph","csi.storage.k8s.io/pv/name":"pvc-05d02323-5e89-4a50-8893-c4ff5840912f","csi.storage.k8s.io/pvc/name":"rbd-pvc","csi.storage.k8s.io/pvc/namespace":"rook-ceph","imageFeatures":"layering","imageFormat":"2","imageName":"csi-vol-3879d00b-1157-11ec-867a-0242ac110005","journalPool":"replicapool","pool":"replicapool","storage.kubernetes.io/csiProvisionerIdentity":"1631180988965-8081-rook-ceph.rbd.csi.ceph.com"},"volume_id":"0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005"}
I0909 11:16:07.743576  140126 cluster_mapping.go:157] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 found new clusterID mapping "rook-ceph1" for existing clusterID "rook-ceph"
I0909 11:16:07.743657  140126 rbd_util.go:1191] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 setting disableInUseChecks: false image features: [layering] mounter: rbd
I0909 11:16:07.755573  140126 omap.go:87] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 got omap values: (pool="replicapool", namespace="", name="csi.volume.3879d00b-1157-11ec-867a-0242ac110005"): map[csi.imageid:1429998da44c csi.imagename:csi-vol-3879d00b-1157-11ec-867a-0242ac110005 csi.volname:pvc-05d02323-5e89-4a50-8893-c4ff5840912f csi.volume.owner:rook-ceph]
E0909 11:16:07.755845  140126 util.go:247] kernel 4.19.202 does not support required features
I0909 11:16:07.887983  140126 cephcmds.go:62] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 command succeeded: rbd [device list --format=json --device-type krbd]
I0909 11:16:07.973578  140126 rbd_attach.go:329] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 rbd: map mon 192.168.61.34:6789
I0909 11:16:08.174316  140126 cephcmds.go:62] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 command succeeded: rbd [--id csi-rbd-node -m 192.168.61.34:6789 --keyfile=***stripped*** map replicapool/csi-vol-3879d00b-1157-11ec-867a-0242ac110005 --device-type krbd --options noudev]
I0909 11:16:08.174341  140126 nodeserver.go:400] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 rbd image: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005/replicapool was successfully mapped at /dev/rbd0
I0909 11:16:08.174427  140126 mount_linux.go:463] Attempting to determine if disk "/dev/rbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/rbd0])
I0909 11:16:08.181456  140126 mount_linux.go:466] Output: ""
I0909 11:16:08.293051  140126 mount_linux.go:463] Attempting to determine if disk "/dev/rbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/rbd0])
I0909 11:16:08.299898  140126 mount_linux.go:466] Output: "DEVNAME=/dev/rbd0\nTYPE=ext4\n"
I0909 11:16:08.299921  140126 mount_linux.go:357] Checking for issues with fsck on disk: /dev/rbd0
I0909 11:16:08.339074  140126 mount_linux.go:453] Attempting to mount disk /dev/rbd0 in ext4 format at /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-05d02323-5e89-4a50-8893-c4ff5840912f/globalmount/0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005
I0909 11:16:08.339116  140126 mount_linux.go:180] Mounting cmd (mount) with arguments (-t ext4 -o _netdev,defaults /dev/rbd0 /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-05d02323-5e89-4a50-8893-c4ff5840912f/globalmount/0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005)
I0909 11:16:08.357826  140126 nodeserver.go:334] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 rbd: successfully mounted volume 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 to stagingTargetPath /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-05d02323-5e89-4a50-8893-c4ff5840912f/globalmount/0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005
I0909 11:16:08.357932  140126 utils.go:188] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 GRPC response: {}
```

The important logs : `I0909 11:16:07.743576  140126 cluster_mapping.go:157] ID: 5 Req-ID: 0001-0009-rook-ceph-0000000000000001-3879d00b-1157-11ec-867a-0242ac110005 found new clusterID mapping "rook-ceph1" for existing clusterID "rook-ceph"`

Works with same cluster id too `I0909 11:44:46.151528  140126 cluster_mapping.go:157] ID: 81 Req-ID: 0001-0009-rook-ceph-0000000000000001-1a77bfb2-1162-11ec-bb4a-0242ac110003 found new clusterID mapping "rook-ceph" for existing clusterID "rook-ceph"`

With above tests, this pr is completely functional and no regression :smile: 
cc @humblec

Rakshith-R commented:
> This pull request now has conflicts with the target branch. Could you please resolve conflicts and force push the corrected changes?

resolved conflicts, PTAL

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.20

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/2438/pipeline/
`etcdserver: request timed out`

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

failed due to #2204 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.22/runs/397/nodes/97/log/?start=0

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.22

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/406/pipeline
failed due to #2264
